wire-runtime/src/commonMain/kotlin/com/squareup/wire/ByteArrayProtoReader32.kt

```diff
@@ -0,0 +1,469 @@
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+package com.squareup.wire
+
+import com.squareup.wire.ProtoReader.Companion.FIELD_ENCODING_MASK
+import com.squareup.wire.ProtoReader.Companion.RECURSION_LIMIT
+import com.squareup.wire.ProtoReader.Companion.STATE_END_GROUP
+import com.squareup.wire.ProtoReader.Companion.STATE_FIXED32
+import com.squareup.wire.ProtoReader.Companion.STATE_FIXED64
+import com.squareup.wire.ProtoReader.Companion.STATE_LENGTH_DELIMITED
+import com.squareup.wire.ProtoReader.Companion.STATE_PACKED_TAG
+import com.squareup.wire.ProtoReader.Companion.STATE_START_GROUP
+import com.squareup.wire.ProtoReader.Companion.STATE_TAG
+import com.squareup.wire.ProtoReader.Companion.STATE_VARINT
+import com.squareup.wire.ProtoReader.Companion.TAG_FIELD_ENCODING_BITS
+import com.squareup.wire.internal.ProtocolException
+import com.squareup.wire.internal.and
+import com.squareup.wire.internal.shl
+import okio.Buffer
+import okio.ByteString
+import okio.ByteString.Companion.toByteString
+import okio.EOFException
+import okio.IOException
+
+internal class ByteArrayProtoReader32(
+  private val source: ByteArray,
+
+  /** The current position in [source], starting at 0 and increasing monotonically. */
+  private var pos: Int = 0,
+
+  /** The absolute position of the end of the current message. */
+  private var limit: Int = source.size,
+) : ProtoReader32 {
+  /** The current number of levels of message nesting. */
+  private var recursionDepth = 0
+
+  /** How to interpret the next read call. */
+  private var state = STATE_LENGTH_DELIMITED
+
+  /** The most recently read tag. Used to make packed values look like regular values. */
+  private var tag = -1
+
+  /** Limit once we complete the current length-delimited value. */
+  private var pushedLimit: Int = -1
+
+  /** The encoding of the next value to be read. */
+  private var nextFieldEncoding: FieldEncoding? = null
+
+  /** Pooled buffers for unknown fields, indexed by [recursionDepth]. */
+  private val bufferStack = mutableListOf<Buffer>()
+
+  /** Lazily-initialized. */
+  private var protoReader: ProtoReader32AsProtoReader? = null
```

Comment:
This variable and the function below are not in ProtoReader. (Should I just use `by lazy` ? I don’t have an intuition on the runtime cost of that)

Reply:
Seems fine as-is